vlog4j-core/src/test/java/org/semanticweb/vlog4j/core/model/AtomImplTest.java

```diff
@@ -29,78 +30,103 @@
+		final Set<Variable> variables = new HashSet<>();
 		variables.add(x);
 		variables.add(y);
 
-		assertEquals("p", atom.getPredicate());
+		assertEquals("p", atom.getPredicate().getName());
+		assertEquals(atom.getTerms().size(), atom.getPredicate().getArity());
+
 		assertEquals(variables, atom.getVariables());
 		assertEquals(Arrays.asList(x, c, d, y), atom.getTerms());
 	}
 
 	@Test
 	public void testEquals() {
-		Variable x = Expressions.makeVariable("X");
-		Constant c = Expressions.makeConstant("c");
-		Atom atom1 = Expressions.makeAtom("p", Arrays.asList(x, c));
-		Atom atom2 = Expressions.makeAtom("p", x, c);
-		Atom atom3 = new AtomImpl("q", Arrays.asList(x, c));
-		Atom atom4 = new AtomImpl("p", Arrays.asList(c, x));
+		final Variable x = Expressions.makeVariable("X");
+		final Constant c = Expressions.makeConstant("c");
+
+		final Predicate predicateP = new PredicateImpl("p", Arrays.asList(x, c).size());
+		final Predicate predicateQ = new PredicateImpl("q", Arrays.asList(x, c).size());
+
+		final Atom atom1 = Expressions.makeAtom("p", Arrays.asList(x, c));
+		final Atom atom2 = Expressions.makeAtom("p", x, c);
+		final Atom atom3 = new AtomImpl(predicateP, Arrays.asList(x, c));
+		final Atom atom4 = new AtomImpl(predicateQ, Arrays.asList(x, c));
+		final Atom atom5 = new AtomImpl(predicateP, Arrays.asList(c, x));
 
 		assertEquals(atom1, atom1);
 		assertEquals(atom1, atom2);
-		assertEquals(atom1.hashCode(), atom2.hashCode());
-		assertNotEquals(atom3, atom2);
-		assertNotEquals(atom3.hashCode(), atom2.hashCode());
-		assertNotEquals(atom4, atom2);
-		assertNotEquals(atom4.hashCode(), atom2.hashCode());
-		assertFalse(atom2.equals(null));
-		assertFalse(atom2.equals(c));
+		assertEquals(atom1, atom3);
+		assertEquals(atom1.hashCode(), atom1.hashCode());
+		assertNotEquals(atom4, atom1);
+		assertNotEquals(atom4.hashCode(), atom1.hashCode());
+		assertNotEquals(atom5, atom1);
+		assertNotEquals(atom5.hashCode(), atom1.hashCode());
+		assertFalse(atom1.equals(null));
+		assertFalse(atom1.equals(c));
 	}
 
 	@Test(expected = NullPointerException.class)
 	public void termsNotNull() {
-		new AtomImpl("p", null);
+		final Predicate predicate1 = Expressions.makePredicate("p", 1);
+		new AtomImpl(predicate1, null);
 	}
 	
 	@Test(expected = IllegalArgumentException.class)
 	public void termsNoNullElements() {
-		Variable x = Expressions.makeVariable("X");
-		new AtomImpl("p",  Arrays.asList(x, null));
+		final Predicate predicate1 = Expressions.makePredicate("p", 1);
+		final Variable x = Expressions.makeVariable("X");
+		new AtomImpl(predicate1, Arrays.asList(x, null));
 	}
 
 	@Test(expected = IllegalArgumentException.class)
-	public void temrsNonEmpty() {
+	public void termsNonEmpty() {
 		Expressions.makeAtom("p");
 	}
 	
 	@Test(expected = NullPointerException.class)
 	public void predicateNotNull() {
-		Expressions.makeAtom(null, Expressions.makeConstant("c"));
+		final Predicate nullPredicate = null;
```

Comment:
what is the use of assigning null to a variable instead of just using it directly?

Reply:
the call  makeAtom(null, Constant constant) is ambiguous, the first argument can be a String (predicate name) or a Predicate